<?xml version="1.0" ?>
<robot name="jeep" xmlns:controller="http://playerstage.sourceforge.net/gazebo/xmlschema/#controller" xmlns:interface="http://playerstage.sourceforge.net/gazebo/xmlschema/#interface" xmlns:sensor="http://playerstage.sourceforge.net/gazebo/xmlschema/#sensor" xmlns:xacro="http://ros.org/wiki/xacro">
  <joint name="jeep/imu_joint" type="fixed">
    <parent link="base_link"/>
    <child link="jeep/imu_link"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <axis xyz="1 0 0"/>
    <limit effort="0.0" lower="0.0" upper="0.0" velocity="0.0"/>
  </joint>
  <joint name="front_left_wheel_joint" type="continuous">
    <parent link="front_left_axle"/>
    <child link="front_left_wheel"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <limit effort="100" velocity="100"/>
    <axis xyz="0 1 0"/>
    <anchor xyz="0 0 0"/>
  </joint>
  <joint name="front_right_wheel_joint" type="continuous">
    <parent link="front_right_axle"/>
    <child link="front_right_wheel"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <limit effort="100" velocity="100"/>
    <axis xyz="0 1 0"/>
    <anchor xyz="0 0 0"/>
  </joint>
  <joint name="rear_left_wheel_joint" type="continuous">
    <parent link="rear_left_axle"/>
    <child link="rear_left_wheel"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <limit effort="100" velocity="100"/>
    <axis xyz="0 1 0"/>
    <anchor xyz="0 0 0"/>
  </joint>
  <joint name="rear_right_wheel_joint" type="continuous">
    <parent link="rear_right_axle"/>
    <child link="rear_right_wheel"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <limit effort="100" velocity="100"/>
    <axis xyz="0 1 0"/>
    <anchor xyz="0 0 0"/>
  </joint>
<!--  <link name="world" />
  <joint name="world_to_base_link=" type="fixed">
    <origin xyz="0 0 1"/>
    <parent link="world"/>
    <child link="base_link"/>
  </joint> -->
  <link name="base_link">
    <inertial>
      <mass value="720.0"/>
      <origin rpy="0 0 0" xyz="0.1 0 0.4"/>
      <inertia ixx="140" ixy="0.0" ixz="0.0" iyy="550" iyz="0.0" izz="550"/>
    </inertial>
    <collision name="base_link_bottom">
      <origin rpy="0 0 0" xyz="-0.28 0 0.405"/>
      <geometry>
        <box size="1.34 1.65746 0.06"/>
      </geometry>
    </collision>
    <collision name="cargo_bottom">
      <origin rpy="0 0 0" xyz="-1.68 0 0.971"/>
      <geometry>
        <box size="1.04609 1.6998 0.01"/>
      </geometry>
    </collision>
    <collision name="cargo_back">
      <origin rpy="0 0 0" xyz="-2.215 0 1.12"/>
      <geometry>
        <box size="0.05 1.69982 0.27"/>
      </geometry>
    </collision>
    <collision name="cargo_left">
      <origin rpy="0 0 0" xyz="-1.705 0.82491 1.12"/>
      <geometry>
        <box size="1.04609 0.05 0.27"/>
      </geometry>
    </collision>
    <collision name="cargo_right">
      <origin rpy="0 0 0" xyz="-1.705 -0.82491 1.12"/>
      <geometry>
        <box size="1.04609 0.05 0.27"/>
      </geometry>
    </collision>
    <collision name="seat">
      <origin rpy="0 0 0" xyz="-0.54 0 0.69"/>
      <geometry>
        <box size="0.52167 1.37206 0.53369"/>
      </geometry>
    </collision>
    <collision name="mud_seat">
      <origin rpy="0 0 0" xyz="-0.54 0 0.93"/>
      <geometry>
        <box size="0.52167 1.30 0.1"/>
      </geometry>
    </collision>
    <collision name="seat_back">
      <origin rpy="0 -0.2 0" xyz="-0.83 0 1.195"/>
      <geometry>
        <box size="0.06 1.37206 0.6"/>
      </geometry>
    </collision>
    <collision name="engine">
      <origin rpy="0 0 0" xyz="0.89 0 0.82"/>
      <geometry>
        <box size="1.4 1.2 0.8"/>
      </geometry>
    </collision>
    <collision name="rollcage_top_front">
      <origin rpy="0 1.23 0" xyz="0.065 0 1.77"/>
      <geometry>
        <box size="0.01299 1.54 0.10226"/>
      </geometry>
    </collision>
    <collision name="rollcage_front_left">
      <origin rpy="0 -0.27 0" xyz="0.112 -0.885 1.51"/>
      <geometry>
        <cylinder length="0.55" radius="0.03"/>
      </geometry>
    </collision>
    <collision name="rollcage_front_right">
      <origin rpy="0 -0.54 0" xyz="0.112 0.885 1.51"/>
      <geometry>
        <cylinder length="0.55" radius="0.03"/>
      </geometry>
    </collision>
    <visual name="visual">
      <origin rpy="0 0 -1.5708" xyz="0 0 0.07"/>
      <geometry>
        <mesh filename="package://targets_path_planning/models/jeep/meshes/body_car.stl" scale="1 1 1"/>
      </geometry>
    </visual>
    <visual name="seat_r">
      <origin rpy="0 0 -1.5708" xyz="0 0 0.07"/>
      <geometry>
        <mesh filename="package://targets_path_planning/models/jeep/meshes/seat_r.stl" scale="1 1 1"/>
      </geometry>
    </visual>
    <visual name="seat_l">
      <origin rpy="0 0 -1.5708" xyz="0 0 0.07"/>
      <geometry>
        <mesh filename="package://targets_path_planning/models/jeep/meshes/seat_l.stl" scale="1 1 1"/>
      </geometry>
    </visual>
    <visual name="spare_wheel">
      <origin rpy="0 0 -1.5708" xyz="0 0 0.07"/>
      <geometry>
        <mesh filename="package://targets_path_planning/models/jeep/meshes/spare_wheel.stl" scale="1 1 1"/>
      </geometry>
    </visual>
    <visual name="steering_wheel">
      <origin rpy="0 0 -1.5708" xyz="0 0 0.07"/>
      <geometry>
        <mesh filename="package://targets_path_planning/models/jeep/meshes/steering_wheel.stl" scale="1 1 1"/>
      </geometry>
    </visual>
    <visual name="front_window">
      <origin rpy="0 0 -1.5708" xyz="0 0 0.07"/>
      <geometry>
        <mesh filename="package://targets_path_planning/models/jeep/meshes/front_window.stl" scale="1 1 1"/>
      </geometry>
    </visual>
  </link>
  <gazebo reference="base_link">
    <material value="Gazebo/Black"/>
  </gazebo>
  <link name="jeep/imu_link">
    <inertial>
      <mass value="0.015"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-05" ixy="0" ixz="0" iyy="1e-05" iyz="0" izz="1e-05"/>
    </inertial>
  </link>

  <link name="front_left_axle">
    <inertial>
      <mass value="0.1"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry name="pioneer_geom">
        <cylinder length="0.01" radius="0.05"/>
      </geometry>
      <material name="AxleGrey">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
  </link>
  <gazebo reference="front_left_axle">
    <material value="Gazebo/Grey"/>
  </gazebo>
  <joint name="base_front_left_axle_joint" type="fixed">
    <origin rpy="0 0 0" xyz="1.09 0.9 0.43"/>
    <parent link="base_link"/>
    <child link="front_left_axle"/>
  </joint>
  <link name="front_right_axle">
    <inertial>
      <mass value="0.1"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry name="pioneer_geom">
        <cylinder length="0.01" radius="0.05"/>
      </geometry>
      <material name="AxleGrey">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
  </link>
  <gazebo reference="front_right_axle">
    <material value="Gazebo/Grey"/>
  </gazebo>
  <joint name="front_right_axle_joint" type="fixed">
    <origin rpy="0 0 0" xyz="1.09 -0.9 0.43"/>
    <parent link="base_link"/>
    <child link="front_right_axle"/>
  </joint>
  <link name="rear_left_axle">
    <inertial>
      <mass value="0.1"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry name="pioneer_geom">
        <cylinder length="0.01" radius="0.05"/>
      </geometry>
      <material name="AxleGrey">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
  </link>
  <gazebo reference="rear_left_axle">
    <material value="Gazebo/Grey"/>
  </gazebo>
  <joint name="rear_left_axle_joint" type="fixed">
    <origin rpy="0 0 0" xyz="-1.653 0.9 0.43"/>
    <parent link="base_link"/>
    <child link="rear_left_axle"/>
  </joint>
  <link name="rear_right_axle">
    <inertial>
      <mass value="0.1"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry name="pioneer_geom">
        <cylinder length="0.01" radius="0.05"/>
      </geometry>
      <material name="AxleGrey">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
  </link>
  <gazebo reference="front_right_axle">
    <material value="Gazebo/Grey"/>
  </gazebo>
  <joint name="rear_right_axle_joint" type="fixed">
    <origin rpy="0 0 0" xyz="-1.653 -0.9 0.43"/>
    <parent link="base_link"/>
    <child link="rear_right_axle"/>
  </joint>
  <link name="front_left_wheel">
    <inertial>
      <mass value="12"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.011763977943"/>
    </inertial>
    <collision name="collision">
      <origin rpy="-1.57079 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.239400" radius="0.417500"/>
      </geometry>
    </collision>
    <visual name="visual">
      <origin rpy="0 0 -1.57079" xyz="1.47 -0.88 -0.31"/>
      <geometry>
        <mesh filename="package://targets_path_planning/models/jeep/meshes/wheel.stl" scale="1.003700111 0.886200464 0.886200464"/>
      </geometry>
    </visual>
  </link>
  <link name="front_right_wheel">
    <inertial>
      <mass value="12"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.011763977943"/>
    </inertial>
    <collision name="collision">
      <origin rpy="-1.57079 0 0" xyz="0 0 0"/> <!-- -0.38277 -0.02 0.11997 -->
      <geometry>
        <cylinder length="0.239400" radius="0.417500"/>
      </geometry>
    </collision>
    <visual name="visual">
      <origin rpy="0 0 1.57079" xyz="-1.47 0.88 -0.31"/>
      <geometry>
        <mesh filename="package://targets_path_planning/models/jeep/meshes/wheel.stl" scale="1.003700111 0.886200464 0.886200464"/>
      </geometry>
    </visual>
  </link>
  <link name="rear_left_wheel">
    <inertial>
      <mass value="12"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.011763977943"/>
    </inertial>
    <collision name="collision">
      <origin rpy="-1.57079 0 0" xyz="0 0 0"/> <!-- -1.47 0.88 0.31 -->
      <geometry>
        <cylinder length="0.239400" radius="0.417500"/>
      </geometry>
    </collision>
    <visual name="visual">
      <origin rpy="0 0 -1.57079" xyz="1.47 -0.88 -0.31"/>
      <geometry>
        <mesh filename="package://targets_path_planning/models/jeep/meshes/wheel.stl" scale="1.003700111 0.886200464 0.886200464"/>
      </geometry>
    </visual>
  </link>
  <link name="rear_right_wheel">
    <inertial>
      <mass value="12"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.011763977943"/>
    </inertial>
    <collision name="collision">
      <origin rpy="-1.57079 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.239400" radius="0.417500"/>
      </geometry>
    </collision>
    <visual name="visual">
      <origin rpy="0 0 1.57079" xyz="-1.47 0.88 -0.31"/>
      <geometry>
        <mesh filename="package://targets_path_planning/models/jeep/meshes/wheel.stl" scale="1.003700111 0.886200464 0.886200464"/>
      </geometry>
    </visual>
  </link>
  <gazebo reference="front_left_wheel">
    <kp>100000000.0</kp>
    <!-- kp and kd for rubber -->
    <kd>1.0</kd>
    <maxContacts>1</maxContacts>
    <fdir1>0 0 1</fdir1>
    <maxVel>1.0</maxVel>
    <mu1>1.0</mu1>
    <mu2>1.0</mu2>
    <minDepth>0.005</minDepth>
    <material value="Gazebo/Black"/>
  </gazebo>
  <gazebo reference="front_right_wheel">
    <kp>100000000.0</kp>
    <!-- kp and kd for rubber -->
    <kd>1.0</kd>
    <maxContacts>1</maxContacts>
    <fdir1>0 0 1</fdir1>
    <maxVel>1.0</maxVel>
    <mu1>1.0</mu1>
    <mu2>1.0</mu2>
    <minDepth>0.005</minDepth>
    <material value="Gazebo/Black"/>
  </gazebo>
  <gazebo reference="rear_left_wheel">
    <kp>100000000.0</kp>
    <!-- kp and kd for rubber -->
    <kd>1.0</kd>
    <maxContacts>1</maxContacts>
    <fdir1>0 0 1</fdir1>
    <maxVel>1.0</maxVel>
    <mu1>1.0</mu1>
    <mu2>1.0</mu2>
    <minDepth>0.005</minDepth>
    <material value="Gazebo/Black"/>
  </gazebo>
  <gazebo reference="rear_right_wheel">
    <kp>100000000.0</kp>
    <!-- kp and kd for rubber -->
    <kd>1.0</kd>
    <fdir1>0 0 1</fdir1>
    <maxContacts>1</maxContacts>
    <maxVel>1.0</maxVel>
    <mu1>1.0</mu1>
    <mu2>1.0</mu2>
    <minDepth>0.005</minDepth>
    <material value="Gazebo/Black"/>
  </gazebo>
  <link name="dir_point">
    <inertial>
      <mass value="0.1"/>
      <origin xyz="0.1 0 0" rpy="0 0 0"/>
      <inertia ixx="140" ixy="0.0" ixz="0.0" iyy="550" iyz="0.0" izz="550"/>
    </inertial>
    <visual>
      <geometry>
        <sphere radius="0.01"/>
      </geometry>
      <material name="white"/>
    </visual>
  </link>
  <gazebo reference="dir_point">
    <material value="Gazebo/White"/>
  </gazebo>
  <joint name="dir_point_joint" type="prismatic">
    <axis xyz="1 0 0"/>
    <limit effort="0" velocity="0" lower="0" upper="0"/>
    <origin rpy="0 0 0" xyz="0.1 0 0"/>
    <parent link="base_link"/>
    <child link="dir_point"/>
  </joint>
  <gazebo>
    <plugin filename="libgazebo_ros_skid_steer_drive.so" name="skid_steer_drive_controller">
      <updateRate>100.0</updateRate>
      <robotNamespace>sim_jeep1</robotNamespace>
      <leftFrontJoint>front_left_wheel_joint</leftFrontJoint>
      <rightFrontJoint>front_right_wheel_joint</rightFrontJoint>
      <leftRearJoint>rear_left_wheel_joint</leftRearJoint>
      <rightRearJoint>rear_right_wheel_joint</rightRearJoint>
      <wheelSeparation>1.8</wheelSeparation>
      <wheelDiameter>0.835</wheelDiameter>
      <robotBaseFrame>base_link</robotBaseFrame>
      <torque>1400</torque>
      <commandTopic>cmd_vel</commandTopic>
      <odometryTopic>odom</odometryTopic>
      <odometryFrame>odom</odometryFrame>
      <broadcastTF>1</broadcastTF>
    </plugin>
  </gazebo>
</robot>
